src/typechecker.erl

```diff
@@ -3490,53 +3495,58 @@ type_check_function(Env, {function,_, Name, NArgs, Clauses}) ->
+add_types_pats([], [], _TEnv, VEnv, PatTysAcc, UBoundsAcc, CsAcc, _Caps) ->
     {lists:reverse(PatTysAcc), lists:reverse(UBoundsAcc),
      VEnv, constraints:combine(CsAcc)};
-add_types_pats([Pat | Pats], [Ty | Tys], TEnv, VEnv, PatTysAcc, UBoundsAcc, CsAcc) ->
+add_types_pats([Pat | Pats], [Ty | Tys], TEnv, VEnv, PatTysAcc, UBoundsAcc, CsAcc, Caps) ->
     NormTy = normalize(Ty, TEnv),
     {PatTyNorm, UBoundNorm, VEnv2, Cs1} =
-        ?throw_orig_type(add_type_pat(Pat, NormTy, TEnv, VEnv),
+        ?throw_orig_type(add_type_pat(Pat, NormTy, TEnv, VEnv, Caps),
                          Ty, NormTy),
     %% De-normalize the returned types if they are the type checked against.
     PatTy  = case PatTyNorm  of NormTy -> Ty;
                                 _      -> PatTyNorm end,
     UBound = case UBoundNorm of NormTy -> Ty;
                                 _      -> UBoundNorm end,
     add_types_pats(Pats, Tys, TEnv, VEnv2,
-                   [PatTy|PatTysAcc], [UBound|UBoundsAcc], [Cs1|CsAcc]).
+                   [PatTy|PatTysAcc], [UBound|UBoundsAcc], [Cs1|CsAcc], Caps).
 
 %% Type check a pattern against a normalized type and add variable bindings.
 -spec add_type_pat(Pat  :: gradualizer_type:abstract_pattern(),
                    Type :: type(),
                    TEnv :: #tenv{},
-                   VEnv :: map()) ->
+                   VEnv :: map(),
+		   Caps :: capture_vars | bind_vars) ->
           {PatTy :: type(), UBound :: type(), NewVEnv :: map(),
            constraints:constraints()}.
-add_type_pat({var, _, '_'}, Ty, _TEnv, VEnv) ->
+add_type_pat({var, _, '_'}, Ty, _TEnv, VEnv, _) ->
     {Ty, Ty, VEnv, constraints:empty()};
-add_type_pat({var, _, A} = Var, Ty, TEnv, VEnv) ->
-    %% TODO: In a fun clause, A is always free, but not in e.g. case clauses
-    case VEnv of
-        #{A := VarTy} ->
-            case subtype(VarTy, Ty, TEnv) of
-                {true, Cs} ->
-                    {type(none), VarTy, VEnv, Cs};
-                false ->
-                    throw({type_error, Var, VarTy, Ty})
-            end;
-        _FreeVar ->
-            {Ty, Ty, VEnv#{A => Ty}, constraints:empty()}
+add_type_pat({var, _, A} = Var, Ty, TEnv, VEnv, CapOrBind) ->
+    case CapOrBind of
+	capture_vars ->
+	    case VEnv of
+		#{A := VarTy} ->
+		    case subtype(VarTy, Ty, TEnv) of
+			{true, Cs} ->
+			    {type(none), VarTy, VEnv, Cs};
+			false ->
+			    throw({type_error, Var, VarTy, Ty})
+		    end;
+		_FreeVar ->
+		    {Ty, Ty, VEnv#{A => Ty}, constraints:empty()}
+	    end;
+	bind_vars ->
+	    {Ty, Ty, VEnv#{A => Ty}, constraints:empty()}
     end;
-add_type_pat(Expr, {type, _, any, []} = Ty, _TEnv, VEnv) ->
+add_type_pat(Expr, {type, _, any, []} = Ty, _TEnv, VEnv, _) ->
     {type(none), Ty, add_any_types_pat(Expr, VEnv), constraints:empty()};
-add_type_pat(Pat, ?type(union, UnionTys)=UnionTy, TEnv, VEnv) ->
+add_type_pat(Pat, ?type(union, UnionTys)=UnionTy, TEnv, VEnv, Cap) ->
```

Comment:
Sometimes it's called Caps and sometimes Cap. It could be consistent. (Not that it matters much though.)

Reply:
Ah, I thought I had fixed that. Clearly I was coding too late at night.